src/main/java/org/opensearch/knn/plugin/script/KNNPainlessScriptUtils.java

```
@@ -0,0 +1,129 @@
+
+            for (List<Number> docVector : docVectors) {
+                if (docVector == null || docVector.isEmpty()) {
+                    continue;
+                }
+
+                float[] dVec = new float[docVector.size()];
+                for (int i = 0; i < docVector.size(); i++) {
+                    dVec[i] = docVector.get(i).floatValue();
+                }
+
+                float similarity = calculateSimilarity(qVec, dVec, space);
+
+                if (isDistanceMetric) {
+                    maxDocTokenSim = Math.min(maxDocTokenSim, similarity);
+                } else {
+                    maxDocTokenSim = Math.max(maxDocTokenSim, similarity);
+                }
+            }
+
+            totalMaxSim += maxDocTokenSim;
+        }
+        return totalMaxSim;
+    }
+
+    private static float calculateSimilarity(@NonNull final float[] vec1, @NonNull final float[] vec2, @NonNull final SpaceType spaceType) {
+        if (vec1.length != vec2.length) {
+            throw new IllegalArgumentException(
+                String.format(
+                    "Vector dimension mismatch in lateInteractionScore: query vector has %d dimensions, document vector has %d dimensions. "
+                        + "Ensure all vectors use the same dimensionality from the same embedding model.",
+                    vec1.length,
+                    vec2.length
+                )
+            );
+        }
+
+        switch (spaceType) {
+            case INNER_PRODUCT:
+                return KNNScoringUtil.innerProduct(vec1, vec2);
+            case COSINESIMIL:
+                return KNNScoringUtil.cosinesimil(vec1, vec2);
+            case L2:
+                return KNNScoringUtil.l2Squared(vec1, vec2);
+            case L1:
+                return KNNScoringUtil.l1Norm(vec1, vec2);
+            case LINF:
+                return KNNScoringUtil.lInfNorm(vec1, vec2);
+            default:
+                return KNNScoringUtil.innerProduct(vec1, vec2);
```

we should throw exception from here rather than defaulting to IP. We have already set the default

---

will throw exception when it's not supported spacetype, also added IT and UT to verify the exception